Fixing Phaser signedness conversion warnings

mijovic opened the pull request:
Part of #9082

mijovic committed:
yul-phaser: A few more tweaks to ensure that signedness conversions do not cause warnings

- Don't cast booleans to int when the result gets assigned to an unsigned type. Relace them with ifs.
- Use unsigned initializer for a variable declared in a lambda capture.
- Unsigned abs() of a difference of unsigned values.

mijovic committed:
yul-phaser: Use ptrdiff_t for iterator arithmetic

axic commented on tools/yulPhaser/Chromosome.cpp:
```diff
@@ -72,5 +72,5 @@ string const& Chromosome::randomOptimisationStep()
 	static vector<string> stepNames = allStepNames();
 
-	return stepNames[SimulationRNG::uniformInt(0, stepNames.size() - 1)];
+	return stepNames[SimulationRNG::uniformInt(0, stepNames.size()) - 1];
```
Is this a conversion warning fix or a bug fix? :wink:

axic replied:
Also wouldn't this result in `0 - 1` when the rng gives 0?

mijovic replied:
This is wrong definitely (in process of changing conversions back and fourth it ended up with this), @cameel wanted to take over this part, so he will fix it I guess :)

cameel replied:
Yeah, looks like a bug. I'll fix it.
BTW, the PR is in the draft state now. I'll mark it ready for review when I'm done with it.

cameel replied:
Fixed.

a3d4 commented on tools/yulPhaser/SimulationRNG.cpp:
```diff
@@ -35,16 +35,16 @@ bool SimulationRNG::bernoulliTrial(double _successProbability)
-uint32_t SimulationRNG::uniformInt(uint32_t _min, uint32_t _max)
+size_t SimulationRNG::uniformInt(size_t _min, size_t _max)
 {
-	boost::random::uniform_int_distribution<> distribution(_min, _max);
```
[`std::uniform_int_distribution`](https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution)? It seems to work with unsigned types too.

mijovic replied:
It has int as template type and this line causes two warnings that unsigned type is converted to signed :)

cameel replied:
@a3d4 Originally I chose boost because I wanted to use `random_device` for seeding and `std::random_device` is not portable (and is known to be implemented in a deterministic way in some environments). I later realized that I don't really need it for my use case (i.e. just a simulation) so I could probably switch to `std::random`. I think I have encountered some other issue that made me stick to `boost` though but I don't really remember now. Do you think switching to `std::random` would be worthwhile?

mijovic replied:
Ah, I did not read what @a3d4 wrote :)
@cameel, it is great if you can swap it with `std::`

cameel replied:
OK. I'll try to swap it then.

cameel replied:
I just realized what my issue with `std::random` was. Basically this: [How can I get an implementation-agnostic version of std::uniform_int_distribution?](https://stackoverflow.com/questions/44520973/how-can-i-get-an-implementation-agnostic-version-of-stduniform-int-distributio)

So using `--seed` option in `yul-phaser` will no longer guarantee reproducible results across different STL implementations. And my earlier experiment results may not be exactly reproducible either (though this might not be all that important since you can easily break reproducibility in other ways, by changing even minor implementation details of the algorithm).

Actually Boost does not have any strong guarantees for this but there's only one implementation so in practice the results will be reproducible (though not necessarily across different Boost versions). That's not the case with STL.

What do you think? Should we switch anyway? Reproducibility is not a hard requirement for phaser to work but it's nice to have. My decision back then was simply "if we depend on Boost anyway, why not use `boost::random`?"

mijovic replied:
@a3d4 if you are fine, I would say(after reading all of this) that we stay with boost::random.

a3d4 replied:
@mijovic 
Sure, if we have good reasons, boost is fine.

> Distribution (both uniform and binomial) doesn't work with unsigned types as there is an error when resolving abs(unsigned)

This still puzzles me. How can I reproduce? Just using `boost::random::uniform_int_distribution<size_t> distribution(_min, _max);` compiles, at least.

cameel replied:
I added a comment in the code explaining why boost is used.

@a3d4
Looks like `uniform_int_distribution<size_t>` works just fine so I'm using that now.

Only binomial has problems. If you change the line to:
```
boost::random::binomial_distribution<size_t> distribution(static_cast<long>(_numTrials), _successProbability);
```
if fails with:
```
/projects/solidity/tools/yulPhaser/SimulationRNG.cpp:55:53:   required from here
/usr/include/boost/random/binomial_distribution.hpp:324:30: error: call of overloaded ‘abs(long unsigned int)’ is ambiguous
```
Apparently the uniform distribution does not use `abs()` and this one does.

a3d4 replied:
OK. BTW, it seems that boost 1.73 is happy with `boost::random::binomial_distribution<size_t>`.

a3d4 replied:
Ah, no, fails with the same error.

a3d4 replied:
Side remark: `boost::random::binomial_distribution<uint16_t>` compiles :).

cameel replied:
Indeed. And `uint32_t` does not. Probably there is only one possible conversion for a 16-bit int so it's not ambiguous.

mijovic replied:
So, decision would be to stay with `boost::random::binomial_distribution<long>`?

cameel replied:
I'm for it. `boost` in this case offers reproducible results. And as for `long` vs `size_t`- phaser never uses the full range anyway and if someone tried, it would be caught by the assertion above.

a3d4 replied:
> Indeed. And uint32_t does not. Probably there is only one possible conversion for a 16-bit int so it's not ambiguous.

I think the issue is quite simple. Forget boost. Consider
```
#include <cmath>
unsigned x = 0;
abs(a);
```
We get the same error, as there is no `abs(unsigned)` (makes some sense), hence few possible overloads.

With `unsigned short` instead of `unsigned` the argument gets promoted to `int` first. (Which is even not guaranteed, `short unsigned` can have the same size as `int` of course.)

cameel committed:
yul-phaser: Switch from uint32_t to size_t in SimulationRNG

- Also pass the appriopriate type internally to the distribution instead of relying on the default (which is uint32_t)

cameel commented:
This is now ready for review.

@mijovic you did a pretty good job with it. I thought that I'd be able to get rid of more but I was only able to improve a few things:
- Fixed the wrong parenthesis (https://github.com/ethereum/solidity/pull/9091#discussion_r433860231)
- Using `0ul` in the lambda eliminated the cast.
- `ClassicGeneticAlgorithm` didn't really require `uint32_t` so I changed it to `size_t`. 
- Eliminated most of the casts in `SimulationRNG` by changing the template arguments.

I put different types of changes in separate commits to make them easier to review. Look at commit messages to see why a particular change was needed.

mijovic commented on tools/yulPhaser/SimulationRNG.cpp:
```diff
@@ -17,34 +17,43 @@
-	boost::random::uniform_int_distribution<> distribution(_min, _max);
-	return distribution(s_generator);
+	boost::random::uniform_int_distribution<size_t> distribution(_min, _max);
+	return static_cast<size_t>(distribution(s_generator));
```
```suggestion
	return distribution(s_generator);
```

cameel replied:
Done.

cameel commented:
Looks like tests were failing because of `bernoulli_distribution<book>`. Should be `bernoulli_distribution<double>`.

Also, the tests for the classic algorithm are really slow. I need to check why.

a3d4 commented on test/yulPhaser/Mutations.cpp:
```diff
@@ -179,8 +179,8 @@ BOOST_AUTO_TEST_CASE(alternativeMutations_should_choose_between_mutations_with_g
 		Chromosome mutatedChromosome = mutation(chromosome);
-		cCount += static_cast<int>(mutatedChromosome == Chromosome("c"));
-		fCount += static_cast<int>(mutatedChromosome == Chromosome("f"));
+		cCount += static_cast<size_t>(mutatedChromosome == Chromosome("c"));
```
int cCount, fCount with no casting?

cameel replied:
Changed to 
```
cCount += (mutatedChromosome == Chromosome("c") ? 1 : 0);
fCount += (mutatedChromosome == Chromosome("f") ? 1 : 0);
```
No cast needed now.

a3d4 commented on test/yulPhaser/TestHelpers.cpp:
```diff
@@ -72,7 +72,7 @@ size_t phaser::test::countDifferences(Chromosome const& _chromosome1, Chromosome
 	size_t count = 0;
 	for (size_t i = 0; i < min(_chromosome1.length(), _chromosome2.length()); ++i)
-		count += static_cast<int>(_chromosome1.optimisationSteps()[i] != _chromosome2.optimisationSteps()[i]);
+		count += static_cast<size_t>(_chromosome1.optimisationSteps()[i] != _chromosome2.optimisationSteps()[i]);
```
```
if (....)
    count++;
```

cameel replied:
Done.

a3d4 commented on test/yulPhaser/TestHelpers.cpp:
```diff
@@ -72,7 +72,7 @@ size_t phaser::test::countDifferences(Chromosome const& _chromosome1, Chromosome
-		count += static_cast<int>(_chromosome1.optimisationSteps()[i] != _chromosome2.optimisationSteps()[i]);
+		count += static_cast<size_t>(_chromosome1.optimisationSteps()[i] != _chromosome2.optimisationSteps()[i]);
 
 	return count + abs(static_cast<int>(_chromosome1.length() - _chromosome2.length()));
```
Unsigned underflow?

cameel replied:
Done.

cameel replied:
In the old code - yeah, but it has been changed in this commit.

But anyway, behavior in this case is well defined. It casts to a signed int properly.

a3d4 commented on tools/yulPhaser/SimulationRNG.cpp:
```diff
@@ -22,29 +22,35 @@
 bool SimulationRNG::bernoulliTrial(double _successProbability)
 {
-	boost::random::bernoulli_distribution<> distribution(_successProbability);
+	boost::random::bernoulli_distribution<double> distribution(_successProbability);
```
`<double>` is default, do you prefer to have it explicit?

mijovic replied:
I would vote for explicit `<double>`. Relying on default can make problem if default changes in different implementation/version.

cameel replied:
I made it explicit because I realized I had wrong assumptions about these types. For example I thought the default was `uint32_t` for the uniform int distribution but it was actually `int` (and I used `uint32_t` for parameters back then not because of it but because the mersenne twister accepts `uint32_t` and I wanted the types consistent - unnecessarily though).

mijovic commented:
@cameel Can you squash commits, please?

cameel commented:
@mijovic I can. But why? I think they add useful information. Each one groups related changes and gives a reason for them.

a3d4 commented:
> I think they add useful information. Each one groups related changes and gives a reason for them.

My personal preference is to have small commits, reflecting the logical steps of implementation. (In particular, I think that having one commit for a comment is perfectly fine.)

But I am not sure if it is in line with solidity repo traditions, which might favor squashing.

mijovic commented:
> > I think they add useful information. Each one groups related changes and gives a reason for them.
> 
> My personal preference is to have small commits, reflecting the logical steps of implementation. (In particular, I think that having one commit for a comment is perfectly fine.)
> 
> But I am not sure if it is in line with solidity repo traditions, which might favor squashing.

I do not have problems with multiple commits, but am used to have one commit per feature as some kind of tradition in solidity repo :)

cameel commented:
As a compromise I squashed some of the commits. There are only three now.

> My personal preference is to have small commits, reflecting the logical steps of implementation. (In particular, I think that having one commit for a comment is perfectly fine.)

Same for me. I always make them very small and only later squash them a little to group it in a way that makes sense. Of course, making sure that each commit is not incomplete, compiles on its own and passes tests. But after that they're usually still pretty small. Apart from grouping changes together it makes it easier to move the code between branches in logical chunks. Or remove part of it if the reviewer requests it.

> I do not have problems with multiple commits, but am used to have one commit per feature as some kind of tradition in solidity repo :)

Well, I've been asked a few times to squash commits but this was mostly when I left them unsquashed just for the sake of review or when I was going back and forth between solutions and wanted to preserve the old solution in history just in case I had to go back to it later. I think this case is different.

Also, one commit rule would be pretty stringent. I often try to make my commits in Solidity a little bigger, but there's usually still more than one logical change per feature.
